Src/DependencyCollector/Shared/Implementation/SqlClientDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,488 @@
+namespace Microsoft.ApplicationInsights.DependencyCollector.Implementation
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Data;
+    using System.Data.SqlClient;
+    using System.Diagnostics;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+    using Microsoft.ApplicationInsights.Web.Implementation;
+
+    internal class SqlClientDiagnosticSourceListener : IObserver<KeyValuePair<string, object>>, IDisposable
+    {
+        // Event ids defined at: https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs
+        public const string DiagnosticListenerName = "SqlClientDiagnosticListener";
+
+        public const string SqlBeforeExecuteCommand = SqlClientPrefix + "WriteCommandBefore";
+        public const string SqlAfterExecuteCommand = SqlClientPrefix + "WriteCommandAfter";
+        public const string SqlErrorExecuteCommand = SqlClientPrefix + "WriteCommandError";
+        
+        public const string SqlBeforeOpenConnection = SqlClientPrefix + "WriteConnectionOpenBefore";
+        public const string SqlAfterOpenConnection = SqlClientPrefix + "WriteConnectionOpenAfter";
+        public const string SqlErrorOpenConnection = SqlClientPrefix + "WriteConnectionOpenError";
+
+        public const string SqlBeforeCloseConnection = SqlClientPrefix + "WriteConnectionCloseBefore";
+        public const string SqlAfterCloseConnection = SqlClientPrefix + "WriteConnectionCloseAfter";
+        public const string SqlErrorCloseConnection = SqlClientPrefix + "WriteConnectionCloseError";
+
+        public const string SqlBeforeCommitTransaction = SqlClientPrefix + "WriteTransactionCommitBefore";
+        public const string SqlAfterCommitTransaction = SqlClientPrefix + "WriteTransactionCommitAfter";
+        public const string SqlErrorCommitTransaction = SqlClientPrefix + "WriteTransactionCommitError";
+
+        public const string SqlBeforeRollbackTransaction = SqlClientPrefix + "WriteTransactionRollbackBefore";
+        public const string SqlAfterRollbackTransaction = SqlClientPrefix + "WriteTransactionRollbackAfter";
+        public const string SqlErrorRollbackTransaction = SqlClientPrefix + "WriteTransactionRollbackError";
+
+        private const string SqlClientPrefix = "System.Data.SqlClient.";
+
+        private readonly TelemetryClient client;
+        private readonly SqlClientDiagnosticSourceSubscriber subscriber;
+
+        private readonly ConcurrentDictionary<Guid, long> operationStartTimestamps = new ConcurrentDictionary<Guid, long>();
```

Comment:
In addition to Liudmila's question: is there any throttling on the diagnostic source so that we may not receive `Error` or `After` and keep increasing memory taken by `operationStartTimestamps`?